### DISCUSSION TITLE
"Dev17"? Stupid question

### DISCUSSION THREAD

**Discussion by DomiOh** [microsoft/STL#1725] [Q&A] [answered]
Hello,

that might be a stupid question, but what is "Dev17" Preview 1?

Greetings.

**Comment by CaseyCarter** [comment, marked as answer]
Yes! Another exciting, "What is the deal with your versioning story?" question!

Microsoft internally refers to each major Visual Studio release as `DevX`, where `X` is a sequential integer value (except for skipping from 12 to 14). Basically, it's the version number that release would have if we had continued with the VC6 versioning scheme instead of switching to versioning by year. Visual Studio 2019 is Dev16, so Dev17 will be the next major release.

**Reply by DomiOh** [reply]
I guessed that to be honest. But I was a bit scared if VS 2019 would be the last VS IDE and will be replaced by something else. Thank you for answering the question. I appreciate it. :)

**Reply by ZaldronGG** [reply]
I guess the only unknown remaining is whether they'll call it Visual Studio 2021 or just drop the number and go evergreen. 😁

**Reply by MikeGitb** [reply]
So the next major version of visual studio is already in the pipeline? Didn't see any official announcements.

**Reply by CaseyCarter** [reply]
> So the next major version of visual studio is already in the pipeline? Didn't see any official announcements.

I can't speculate or comment on future products or release dates, etc., but I can confirm that 16 + 1 is 17 =)

**Reply by jovibor** [reply]
It's quite sane to assume that the next major VS release will be timed exactly to the full `C++20` compiler/library support.
Which is, as I can guess, approximately 2-3 months  from now.

The more interesting question is when the [`vNext`](https://github.com/microsoft/STL/issues?q=is%3Aopen+is%3Aissue+label%3AvNext) version will see the light.

**Reply by ZaldronGG** [reply]
> The more interesting question is when the [`vNext`](https://github.com/microsoft/STL/issues?q=is%3Aopen+is%3Aissue+label%3AvNext) version will see the light.

FWIW, last time (Dev16) there were five months between Preview 1 and Release.

**Reply by StephanTLavavej** [reply]
At this point, our plans for the "vNext" ABI-breaking release are still unformed (there's a [wiki page](https://github.com/microsoft/STL/wiki/vNext-Planning) with various thoughts), but to avoid any confusion, it is essentially certain that the "Dev17" 17.0 release will *not* coincide with vNext. Instead, it is highly likely that 17.0 will remain binary-compatible (with VS 2015/2017/2019), and in some future update 17.x, the "vNext" ABI-breaking libraries will be an *option* supported alongside the binary-compatible libraries. We're currently thinking that there will be a feature switchover - we'll finish C++20 in the binary-compatible release series, and then all C++23 work will go into vNext. (Doing feature work in two radically different branches is way too much work, as we learned from previous experience.)

**Reply by jovibor** [reply]
> ...it is highly likely that 17.0 will remain binary-compatible (with VS 2015/2017/2019), and in some future update 17.x, the "vNext" ABI-breaking libraries will be an _option_ supported alongside the binary-compatible libraries. 

Thanks for the thorough explanation.

Believe it or not, decades-long ABI compatibility is the worst thing that exists in the C++ world, even worse than implicit conversions.   

Every sane lib you are about to use should be available for you in a source-code, so you can always recompile it. But instead we must live in this ABI-compatibility world that every person from committee and big vendors is so afraid to throw away.

**Reply by MikeGitb** [reply]
> I can't speculate or comment on future products or release dates, etc., but I can confirm that 16 + 1 is 17 =)

Not expecting a confirmation, but my comment was based on the fact that in https://github.com/microsoft/STL/wiki/Changelog, Dev17 is listed directly after 16.10 instead of a hypothetical 16.11.

@StephanTLavavej : If you go that route (it sounds like a good idea to me), please think really hard about a good naming scheme to distinguish the two ABI versions. IIRC, ABI breaking changes where always aligned to new versions of Visual studio, which made it easy to specify for which version a particular library was compiled (e.g. qt_core_msvc_2013, qt_core_msvc_2015). 

Also there is the "danger" of this becomming unintentiionally mixed up with the c++ version as in "the c++20 vs c++23 abi  even though my guess is that you could use the new standard library ABI, but still compile your code in c++14/17/20 mode (although it might be interesting to see how much code could be removed if the new version only had to support c++23 and up)

**Reply by miscco** [reply]
I am super terrified by the C++20 / C++23 switch. 

Looking at the desire to redesign certain containers this dual track wil become a maintenance hell (Except the vNext branch moves to its own files somehow)

**Reply by StephanTLavavej** [reply]
@jovibor We're very familiar with the downsides of ABI compatibility, as it has prevented us from fixing major things (which is why we're going to have vNext). I was around back when we broke ABI every major version (the full release cycles of VS 2010, 2012, 2013, and 2015) and it was glorious for implementers - we got to fix tons of stuff and rip out tons of dead code. However, customers found this hard to keep up with - regardless of how in an ideal world, everyone should be able to rebuild everything from source on demand. This led to delayed upgrades and ultimately harmed progress (having a really cool new major version doesn't do any good if most people are still stuck on VS 2010 or whatever). Maintaining compatibility since VS 2015 has been very successful - but it is clear that we do need to break ABI from time to time. My hope is that we'll be able to establish a semi-predictable pattern, offering stability for multiple releases/5+ years, but establishing the expectation that ABI breaks and migration will be periodically necessary.

@MikeGitb Yes, we'll definitely think about naming to mitigate confusion as much as possible. I hope that nobody refers to "the C++23 ABI" since that will be confusing - vNext will certainly support C++20 and C++23 as distinct modes, and is virtually certain to support C++17. (I am holding out hope that we may be able to drop C++14 mode, and I want a pony and a unicorn. Perhaps I will get one.)

@miscco The idea of a clean switchover is that there will be no maintenance of the old version - only critical fixes (much as we already do today for the VS 2015 and VS 2017 toolsets, which are still supported but not with active development). It's not certain, but I expect that GitHub's history will remain linear, and we'll simply say that `main` is a vNext branch when the switchover happens. Internally, we'll have separate branches for the "v19" compatible and vNext toolsets, but the old branch will be rarely changed. This is precisely because we tried to simultaneously develop two branches for "classic vNext" 3 years ago and it became unmaintainable.

**Reply by MikeGitb** [reply]
> and I want a pony and a unicorn. Perhaps I will get one.)

🐴🦄 (sorry couldn't resist). 

Thanks for sharing your thoughts on this and lets hope for the best.

**Reply by MikeGitb** [reply]
Any news on vnext?

**Reply by StephanTLavavej** [reply]
No vNext news, sorry.
